File: web/src/lib/debtService.ts
import { db } from './db';
import { orderDebt } from '@/types';

export interface DebtEntry {
  customerId: number;
  name: string;
  phone: string | null;
  totalDebt: number;
  orderCount: number;
  oldestOrderAt: number | null;
}

export async function getDebtList(): Promise<DebtEntry[]> {
  const [orders, customers] = await Promise.all([
    db.orders.toArray(),
    db.customers.toArray(),
  ]);
  const customerById = new Map(customers.map((c) => [c.id!, c]));
  const byCustomer = new Map<
    number,
    { debt: number; orderCount: number; oldestAt: number | null }
  >();

  for (const o of orders) {
    const d = orderDebt(o);
    if (d <= 0) continue;
    const cur = byCustomer.get(o.customer_id) ?? {
      debt: 0,
      orderCount: 0,
      oldestAt: null,
    };
    cur.debt += d;
    cur.orderCount += 1;
    cur.oldestAt =
      cur.oldestAt == null ? o.created_at : Math.min(cur.oldestAt, o.created_at);
    byCustomer.set(o.customer_id, cur);
  }

  const list: DebtEntry[] = [];
  for (const [customerId, data] of byCustomer) {
    const c = customerById.get(customerId);
    if (!c) continue;
    list.push({
      customerId,
      name: c.name,
      phone: c.phone ?? null,
      totalDebt: data.debt,
      orderCount: data.orderCount,
      oldestOrderAt: data.oldestAt,
    });
  }

  list.sort((a, b) => b.totalDebt - a.totalDebt);
  return list;
}

export async function getTotalDebt(): Promise<number> {
  const list = await getDebtList();
  return list.reduce((sum, e) => sum + e.totalDebt, 0);
}

File: app/lib/services/debt_service.dart
import '../database/database_helper.dart';
import '../models/order.dart';
import 'customer_service.dart';

class DebtEntry {
  final int customerId;
  final String name;
  final String? phone;
  final double totalDebt;
  final int orderCount;
  final int? oldestOrderAt;

  const DebtEntry({
    required this.customerId,
    required this.name,
    this.phone,
    required this.totalDebt,
    required this.orderCount,
    this.oldestOrderAt,
  });
}

class DebtService {
  final DatabaseHelper _db = DatabaseHelper();
  final CustomerService _customers = CustomerService();

  Future<List<DebtEntry>> getDebtList() async {
    final db = await _db.database;
    final orderRows = await db.query('orders');
    final allCustomers = await _customers.getAllCustomers();
    final byId = {for (final c in allCustomers) c.id!: c};

    final agg = <int, ({double debt, int orderCount, int? oldestAt})>{};
    for (final r in orderRows) {
      final o = Order.fromMap(r);
      final d = o.debt;
      if (d <= 0) continue;
      final cur = agg[o.customerId];
      if (cur == null) {
        agg[o.customerId] = (
          debt: d,
          orderCount: 1,
          oldestAt: o.createdAt,
        );
      } else {
        agg[o.customerId] = (
          debt: cur.debt + d,
          orderCount: cur.orderCount + 1,
          oldestAt: cur.oldestAt == null
              ? o.createdAt
              : (o.createdAt < cur.oldestAt! ? o.createdAt : cur.oldestAt),
        );
      }
    }

    final list = <DebtEntry>[];
    for (final e in agg.entries) {
      final c = byId[e.key];
      if (c == null) continue;
      list.add(DebtEntry(
        customerId: e.key,
        name: c.name,
        phone: c.phone,
        totalDebt: e.value.debt,
        orderCount: e.value.orderCount,
        oldestOrderAt: e.value.oldestAt,
      ));
    }
    list.sort((a, b) => b.totalDebt.compareTo(a.totalDebt));
    return list;
  }

  Future<double> getTotalDebt() async {
    final list = await getDebtList();
    return list.fold<double>(0.0, (sum, e) => sum + e.totalDebt);
  }
}
